### ISSUE TITLE
关于源码中Lreg损失缺失的问题

### ISSUE THREAD

**Issue by mbhnwz** [Muyiiiii/NeurIPS-25-Glocal-IB#2]
你好：
       我最近在研究时序填补的技术并发现了您的论文。 我在阅读源码过程中，发现源码和论文中存在部分不一致，具体为：Lreg损失（KL散度）在源码中缺失，仅保留了MSE损失和Lglo。
       请问这是有意为之的设计，还是遗漏。
 
       期待您的答复，多谢。

**Comment by Muyiiiii** [comment]
Hi mbhnwz,

Thank you for your interest in our work and for taking the time to examine the source code so closely. We appreciate your detailed observation regarding the discrepancy between the paper and the implementation.

To answer your question: the exclusion of the KL divergence loss ($L_\text{reg}$) in the released code was an **intentional design choice** to prioritize usability and flexibility. Here are the main reasons behind this decision:

1. **Plug-and-Play Flexibility:** Our primary goal was to design the method as a "plug-and-play" module that can be seamlessly inserted into existing Time Series Imputation models. Strictly implementing the KL loss requires the backbone model's intermediate layer to utilize a variational structure (outputting both $\mu$ and $\sigma$ for reparameterization). Since many standard backbones do not support this natively, removing the strict KL requirement allows our method to be compatible with a wider range of architectures without modifying their core design.

2. **Empirical Performance:** Our extended experiments across various models demonstrated that the method maintains robust performance even without the $L_\text{reg}$ term.

3. **Data Scale & Ablation:** As shown in our ablation studies, given the relatively small scale of many time-series datasets, the regularization benefit provided by the KL loss is often limited. Therefore, we simplified the implementation to retain efficiency without sacrificing significant accuracy.

4. **Implementation of KL loss**:

   ```python
   import torch
   import torch.nn as nn
   
   # Assume hidden_features comes from your backbone model
   hidden_features = torch.randn(32, 64) # (Batch, Hidden_Dim)
   hidden_dim = 64
   latent_dim = 16
   Duplicate of #
   # 1. Structural Requirement: Two separate predictors for mu and log_var
   head_mu = nn.Linear(hidden_dim, latent_dim)
   head_logvar = nn.Linear(hidden_dim, latent_dim)
   
   # 2. Generate Parameters
   mu = head_mu(hidden_features)
   log_var = head_logvar(hidden_features) # Estimates log(sigma^2)
   
   # 3. Calculate KL Divergence Loss
   # Formula: -0.5 * sum(1 + log(sigma^2) - mu^2 - sigma^2)
   kl_loss = -0.5 * torch.sum(1 + log_var - mu.pow(2) - log_var.exp())
   
   print(f"KL Loss: {kl_loss.item():.4f}")
   ```

   

I hope this clarifies the design rationale. Please feel free to reach out if you have any further questions.

Best regards,
Authors of Glocal-IB

**Comment by mbhnwz** [comment]
你好，
感谢您的回复

1、您的意思是模型编码器输出的后验分布需要满足高斯分布，但是这个高斯分布的均值和方差需要模型单独的结构预测出来？目前看到的源码中貌似没有模型会有该项操作。 同时，hidden_features = torch.randn(32, 64)为什么是两维的，根据代码它通常是(batch_size, seq_len, model_dim)三维结构。
2、在阅读代码中ContrastiveLoss计算中，有一项新的小疑问，具体为：
以ETTh1数据集为例，MyContrastiveLoss_v2中计算损失时，results["X_obs_p_contras"]和results["X_ori_z_contras"]的size为(batch_size, seq_len, model_dim)，即(32.96,256)。
为什么计算余弦相似度 loss_calc_func: Criterion = nn.CosineSimilarity(eps=1e-8, dim=1)，会在dim=1，也就是时间维进行计算，而非在model_dim维度。

期待您的答复，多谢。

**Comment by Muyiiiii** [comment]
Hi mbhnwz,

Thank you for your follow-up questions. Please find my explanations below:

1. Regarding the KL Loss and Model Structure

   Yes, you are correct. To incorporate the KL loss, the model’s embedding layer must explicitly predict the parameters of the Gaussian distribution. Specifically, you can treat the predicted mean ($\mu$) as the input for the subsequent decoder. For the implementation details of this mechanism, I recommend referring to the reparameterization trick and KL divergence calculation found in the original VAE paper.

   Regarding your observation about the dimensions: the `hidden_features` used in the previous snippet was just a simplified example. In the actual training code, the tensor structure indeed follows the standard 3D shape, typically `(batch, n_channel, seq_len)`.

2. Regarding the Contrastive Loss Dimension (dim=1)

   The decision to apply Cosine Similarity along dim=1 (the temporal dimension) is intentional.

   **Reasoning:** By calculating similarity over the sequence length, we treat the temporal trajectory of each feature as a vector. This enforces **Temporal Consistency**, encouraging the model to match the **shape and trend** (dynamics) of the time series across the entire window, rather than just aligning point-wise features at individual timestamps (which would correspond to `dim=2`).

   In the context of imputation, capturing the correct temporal evolution (rise and fall patterns) is often more critical than matching instantaneous feature embeddings, which is already effectively handled by the complementary MSE loss.

   **But it just depends on your tasks.**

I hope this helps clarify your questions.
